### PR TITLE
Improvements on reassembling back annotations from InfluxDB.

### PR DESCRIPTION
**Pull request by chris-ramon**
#### Details

Issue: #117 

- [x] tl;dr - Improves how we reassemble InfluxDB fields to span annotations:
  - Updates `func annotationsFromRow(...)` to include common logic that we use to reassemble data back from InfluxDB fields to span `annotations`. Now it uses `func annotationsFromEvents(..)` which will discard those annotations that should no be included because were not explicitly save by `Collect(...)` but that are present because `InfluxDB` will return all the fields present on `spans` measurement:
    - Eg. A field named `Client.Request.Method` might contain values for sub-spans but not for root-spans. Since span annotations are non-deterministic, current strategy for query those annotations(fields) is to do `select * from spans ....`, therefore: `Client.Request.Method` will be included in the query result even for root-spans:
screen shot 2016-03-17 at 14 18 14

### CODE REVIEW

**Comment by chris-ramon** [comment]
tl:dr —There's not performance impact introduced to the `InfluxDBStore.Trace(...) & InfluxDBStre.Traces(...)` with this PR.

I went ahead and re-run benchmarks from #114 for `InfluxDBStore.Trace(...)` & `InfluxDBStore.Traces(...)` in order to confirm if there are big performance impact with [changes](https://github.com/sourcegraph/appdash/pull/121/files#diff-b7da646763ca1f5622f8d108c536ce86R422) introduced in this PR because unmarshalling/marshalling via: [`UnmarshalEvents`](https://github.com/sourcegraph/appdash/blob/955366b27d4851653681b18179a6b905f932f2ed/event.go#L134) & [`MarshalEvent`](https://github.com/sourcegraph/appdash/blob/955366b27d4851653681b18179a6b905f932f2ed/event.go#L42).

- `InfluxDBStore.Trace(...)`

  - Before:

```bash
$ go test -run=XYZ -bench BenchmarkInfluxDBStoreTrace. -benchtime=40s > trace
$ more trace

PASS
BenchmarkInfluxDBStoreTrace100-4             200         411849069 ns/op
BenchmarkInfluxDBStoreTrace250-4              50        1051788921 ns/op
BenchmarkInfluxDBStoreTrace1000-4             10        4227713513 ns/op
ok      sourcegraph.com/sourcegraph/appdash     380.156s
```

  - After:

```bash
$ go test -run=XYZ -bench BenchmarkInfluxDBStoreTrace[0-9]+ -benchtime=40s > trace
$ more trace

PASS
BenchmarkInfluxDBStoreTrace100-4             100         412541990 ns/op
BenchmarkInfluxDBStoreTrace250-4              50        1079761891 ns/op
BenchmarkInfluxDBStoreTrace1000-4             10        4208731896 ns/op
ok      sourcegraph.com/sourcegraph/appdash     294.010s
```

- `InfluxDBStore.Traces(...)`

Before:

```bash
$ go test -run=XYZ -bench BenchmarkInfluxDBStoreTracesDefaultPerPage > traces
$ more traces

PASS
BenchmarkInfluxDBStoreTracesDefaultPerPage-4         100          13433622 ns/op
ok      sourcegraph.com/sourcegraph/appdash     10.595s
```

After:

```bash
$ go test -run=XYZ -bench BenchmarkInfluxDBStoreTracesDefaultPerPage > traces
$ more traces
PASS
BenchmarkInfluxDBStoreTracesDefaultPerPage-4         100          13300171 ns/op
ok      sourcegraph.com/sourcegraph/appdash     11.540s
```

File: event_test.go [hunk 1]
```diff
@@ -83,7 +83,10 @@ func TestUnmarshalEvent(t *testing.T) {
 }
 
 func TestUnmarshalEvents(t *testing.T) {
-	origRegisteredEvents := registeredEvents
+	origRegisteredEvents := map[string]Event{}
```
Comment (neelance) [inline event_test.go]: While this was certainly a bug (good catch!), I would solve it differently: In line 91-93 (original) the map gets cleared entirely. So we can just do:
```
origRegisteredEvents := registeredEvents
defer func() {
	registeredEvents = origRegisteredEvents
}()
registeredEvents = make(map[string]Event)
```

Comment (chris-ramon) [inline event_test.go]: Thanks for reviewing @neelance, good call on the shorter/[performant](https://groups.google.com/forum/#!msg/golang-nuts/UvUm3LA1u8g/jGv_FobNpN0J) approach to clear map; commit 70d8733e8273884ad6631df840c36f22df3a82bf includes the change.

Comment (neelance) [inline event_test.go]: You're still copying the map. ;-)

Comment (chris-ramon) [inline event_test.go]: Right! not need to copy, anymore - thanks!

**Comment by slimsag** [comment]
I hope to test this change with our app soon to confirm, but this looks like a very good change. This seems much more correct than #119 and using using MarshalEvent/UnmarshalEvent seems correct to me (but also highlights why Appdash should eventually move to a more simplified storage modal).

I'll follow-up with more issues if I spot them in my testing. Thanks Chris! :+1:
